File: CS/XtraPivotGrid_CustomFilterItemsSorting/Form1.cs
using System;
using System.Collections;
using System.Windows.Forms;
using DevExpress.XtraPivotGrid;
using DevExpress.XtraPivotGrid.Data;

namespace XtraPivotGrid_SortedFilterItems
{
    public partial class Form1 : DevExpress.XtraEditors.XtraForm
    {
        public Form1() {
            InitializeComponent();
            excelDataSource1.FileName = "SalesPerson.xlsx";
            excelDataSource1.Fill();
            pivotGridControl1.CustomFilterPopupItems += pivotGridControl1_CustomFilterPopupItems;
        }
        private void Form1_Load(object sender, EventArgs e) {
            pivotGridControl1.DataSource = excelDataSource1;
            pivotGridControl1.BestFit();
        }
        private void pivotGridControl1_CustomFilterPopupItems(object sender, 
            PivotCustomFilterPopupItemsEventArgs e) {
            if(rbCaptionLength.Checked)
                ArrayList.Adapter((IList)e.Items).Sort(new FilterItemsComparer());
        }
        private void rbAlphabeticalSort_CheckedChanged(object sender, EventArgs e) {
            fieldProductName1.ShowFilterPopup();
        }
    }
    public class FilterItemsComparer : IComparer {
        int IComparer.Compare(object x, object y) {
            if (!(x is PivotGridFilterItem) || !(y is PivotGridFilterItem)) return 0;
            PivotGridFilterItem item1 = (PivotGridFilterItem)x;
            PivotGridFilterItem item2 = (PivotGridFilterItem)y;
            if (item1.ToString().Length == item2.ToString().Length) return 0;
            if (item1.ToString().Length > item2.ToString().Length) return 1;
            return -1;
        }
    }
}

File: VB/XtraPivotGrid_CustomFilterItemsSorting/Form1.vb
Imports System
Imports System.Collections
Imports System.Windows.Forms
Imports DevExpress.XtraPivotGrid
Imports DevExpress.XtraPivotGrid.Data

Namespace XtraPivotGrid_ExampleTemplate
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
			AddHandler pivotGridControl1.CustomFilterPopupItems, AddressOf pivotGridControl1_CustomFilterPopupItems
		End Sub
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
			Dim tableAdapter As New DataSet1TableAdapters.ProductReportsTableAdapter()
			pivotGridControl1.DataSource = tableAdapter.GetData()
			pivotGridControl1.BestFit()
		End Sub
		Private Sub pivotGridControl1_CustomFilterPopupItems(ByVal sender As Object, ByVal e As PivotCustomFilterPopupItemsEventArgs)
			If rbCaptionLength.Checked Then
				ArrayList.Adapter(DirectCast(e.Items, IList)).Sort(New FilterItemsComparer())
			End If
		End Sub
		Private Sub rbAlphabeticalSort_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rbAlphabeticalSort.CheckedChanged
			fieldProductName.ShowFilterPopup()
		End Sub
	End Class
	Public Class FilterItemsComparer
		Implements IComparer

		Private Function IComparer_Compare(ByVal x As Object, ByVal y As Object) As Integer Implements IComparer.Compare
			If Not (TypeOf x Is PivotGridFilterItem) OrElse Not (TypeOf y Is PivotGridFilterItem) Then
				Return 0
			End If
			Dim item1 As PivotGridFilterItem = DirectCast(x, PivotGridFilterItem)
			Dim item2 As PivotGridFilterItem = DirectCast(y, PivotGridFilterItem)
			If item1.ToString().Length = item2.ToString().Length Then
				Return 0
			End If
			If item1.ToString().Length > item2.ToString().Length Then
				Return 1
			End If
			Return -1
		End Function
	End Class
End Namespace